proposals/000-cooperative-matrix-2.md

```diff
@@ -0,0 +1,536 @@
+    CoopMatTensorView<DimView, HasDimensions> tensorView,
+    FuncType decodeFunc);
+```
+
+`decodeFunc` is a user-defined function that allows to decode if it was encoded.
+ - A function (or functor) whose return type matches the matrix element type.
+ - The first parameter is a pointer to the raw buffer data for the element.
+ - The second parameter is a uint32_t[Dim] array for block coordinates.
+ - The third parameter is a uint32_t[Dim] array for coordinates within the block.
+ - All parameters are read-only.
+ - The function is called for each matrix element, and its return value is stored in the result.
+
+An example of `decodeFunc` will be,
+```
+T decodeFunc(
+    const T* ptr,  // pointer to the raw data for the element
+    const uint32_t[Dim] blockCoord,   // block coordinates
+    const uint32_t[Dim] coordInBlock  // coordinates within the block
+);
+```
+
+**TODO**: It is unclear if we can use a raw memory pointer for the first parameter.
+
+```
+__generic<
+    let Dim : uint32_t,
+    let ClampMode : CoopMatClampMode>
+void store(
+    AnyRWBuffer buf,
+    uint elementOffset,
+    CoopMatTensorLayout<Dim, ClampMode> tensorLayout);
+
+__generic<
+    let Dim : uint32_t,
+    let ClampMode : CoopMatClampMode,
+    let DimView : uint32_t,
+    let HasDimensions : bool>
+void store(
+    AnyRWBuffer buf,
+    uint elementOffset,
+    CoopMatTensorLayout<Dim, ClampMode> tensorLayout,
+    CoopMatTensorView<DimView, HasDimensions> tensorView);
+```
+
+### Matrix Multiply-and-Accumulate operation
+
+`coopMatMulAdd()` performs a linear-algebraic matrix multiply of A by B and then component-wise add C.
+
+A, B, C and result matrices must use a same scope.
+
+The component types of A, B, C, result matrices don't need to be same.
+
+```
+__generic<
+    T : __BuiltinArithmeticType,
+    U : __BuiltinArithmeticType,
+    V : __BuiltinArithmeticType,
+    W : __BuiltinArithmeticType,
+    let S : CoopMatScope,
+    let M : int,
+    let K : int,
+    let N : int>
+CoopMat<T, S, M, N, CoopMatMatrixUse::MatrixAccumulator> coopMatMulAdd(
+    CoopMat<U, S, M, K, CoopMatMatrixUse::MatrixA> matA,
+    CoopMat<V, S, K, N, CoopMatMatrixUse::MatrixB> matB,
+    CoopMat<W, S, M, N, CoopMatMatrixUse::MatrixAccumulator> matC,
+    constexpr CoopMatMatrixOperands operands = CoopMatMatrixOperands::None);
+```
+
+### Reduce operation
+
+`coopMatReduce()` computes a matrix where each element of the result matrix is computed from a row, column, or neighborhood of the source matrix depending on `reduceMask`.
+
+```
+__generic<
+    T : __BuiltinFloatingPointType,
+    let S : CoopMatScope,
+    let M : int,
+    let N : int,
+    let RM : int,
+    let RN : int,
+    FuncType : IFunc<T, T, T>>
+void coopMatReduce(
+    out CoopMat<T, S, RM, RN, CoopMatMatrixUse::MatrixAccumulator> result,
+    CoopMat<T, S, M, N, CoopMatMatrixUse::MatrixAccumulator> m,
+    CoopMatReduceMask reduceMask,
+    FuncType combineOp);
```

Comment:
this should be `functype(T,T)->T combineOp` because SPIRV doesn't allow `this` in a reduce function.